feat: improve block_chain backend reliability

SWvheerden opened the pull request:
Description
---
This PR tries to improve db transactions to make sure all changes are atomic and can be reset if need be, also improves logging and comments. 

Motivation and Context
---
We changed the rewind logic to use smaller multiple transactions after we encountered a few too-large transactions. 
Between the too-large transactions and using multiple transactions, if they fail, they might leave the database in an inconsistent state. This PR tries to ensure when adding information, we always use a single transaction and never leave the database in an inconsistent state. 

This PR changes the rewind to try make sure that if it fails, the rewind will try to reset it self back to the previous state. 

How Has This Been Tested?
---

- Manually sync node, and run node
- Unit tests


Fixes: #4850

SWvheerden committed:
improve black_chain backend reliability

stringhandler commented:
I'd be happier if there was a unit test showing that this works

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1771,6 +1768,46 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+    let mut txn = DbTransaction::new();
+    info!(target: LOG_TARGET, "Deleting block {}", height);
+    let block = fetch_block(db, height, false)?;
+    debug!(target: LOG_TARGET, "Goint to delete block: {}", block);
```
```suggestion
    debug!(target: LOG_TARGET, "Going to delete block: {}", block);
```

stringhandler commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1985,11 +2023,19 @@ fn reorganize_chain<T: BlockchainBackend>(
+        if let Err(e) = insert_best_block(&mut txn, block.clone()) {
+            warn!(
+                target: LOG_TARGET,
+                "Failed extract Monero data: {:?}. Restoring last chain.", e
```
This doesn't look like the right error. Is it correct?

SWvheerden replied:
Yeah it is correct. 
Insert best block can only fail if it fails to deserialise the monero data.

sdbondi replied:
nit: That is an internal detail of insert_best_block though, so the warn log text should be agnostic to that. The `e` can/should describe the problem itself.

stringhandler reviewed (commented):
I think these changes are good, but I can't merge this without tests proving the case

stringhandler commented on base_layer/core/tests/chain_storage_tests/chain_storage.rs:
```diff
@@ -238,6 +238,41 @@ fn test_rewind_to_height() {
+    let network = Network::LocalNet;
+
+    let mut db = TempDatabase::new();
+    db.set_fail_write_counter(12);
```
need a comment with the significance of 12 here

SWvheerden committed:
remove redundant db query

clippy

Update base_layer/core/src/chain_storage/blockchain_database.rs

Co-authored-by: stringhandler <stringhandler@gmail.com>
Auto stash before checking out "SWvheerden/sw_improve_backend"

add comment about 12

sdbondi commented on base_layer/core/src/chain_storage/blockchain_database.rs:
```diff
@@ -1711,46 +1713,41 @@ fn rewind_to_height<T: BlockchainBackend>(db: &mut T, height: u64) -> Result<Vec
+                );
+                let mut txn = DbTransaction::new();
+
+                for block in removed_blocks.into_iter().rev() {
```
Wouldn't you call `restore_reorged_chain` here?

sdbondi replied:
Ah i see that function calls rewind_to_hash

SWvheerden commented:
I am going to close this for now, this is way outdated. This was an attempt at improving reliability when we had a bug in the rewind code. 
This makes many small improvements in the layout of the code. 
BUT this touches core critical code that at this moment works perfectly. I am not in favor of touching working code.
